brain-marks/Services/MigrationService.swift

```diff
@@ -0,0 +1,90 @@
+//
+//  MigrationService.swift
+//  brain-marks
+//
+//  Created by Jay on 12/13/22.
+//
+
+import CoreData
+import Foundation
+
+/// Controls the migration from Amplify to CoreData
+class MigrationService {
+    private let storageProvider = StorageProvider.shared
+    private var awsCategories: [AWSCategory] = []
+
+    private let amplifyDataStore = DataStoreManger.shared
+
+    private var managedObjectContext: NSManagedObjectContext
+
+    init() {
+        self.managedObjectContext = storageProvider.context
+    }
+
+    func performMigration() {
+        /// 1. Fetch categories
+        getCategories()
+        /// 2. Go through each category and get tweets
+        for awsCategory in awsCategories {
+            /// 3. Add category to core data
+            let categoryToAdd = makeCategoryCoreDataCompatible(category: awsCategory)
+            amplifyDataStore.fetchSavedTweets(for: awsCategory) { awsTweets in
+                for tweet in awsTweets ?? [] {
+                    let tweetToAdd = self.makeTweetCoreDataCompatible(tweet: tweet)
+                    categoryToAdd.addToTweets(tweetToAdd)
+                }
+            }
+            do {
+                try managedObjectContext.save()
+            } catch {
+                print("❌ MigrationService.performMigration() Error: \(error)")
+            }
+        }
+        UserDefaults.standard.set(true, forKey: "migrationToCoreDataRan")
+    }
+
+    public func checkIfMigrationShouldRun() -> Bool {
+        let migrationToCoreDataRan = UserDefaults.standard.bool(forKey: "migrationToCoreDataRan")
+
+        if storageProvider.getAllCategories().isEmpty && !migrationToCoreDataRan {
+            return true
+        }
+        return false
+    }
+
+    private func getCategories() {
+        amplifyDataStore.fetchCategories(completion: { result in
+            switch result {
+            case .success(let categories):
+                self.awsCategories = categories
+            case .failure(let error):
+                print("❌ MigrationService.getCategories(): Error: \(error)")
+            }
+        })
+    }
+
+    private func makeCategoryCoreDataCompatible(category: AWSCategory) -> CategoryEntity {
```

Comment:
Naming is hard. I'm down to rename this but I'm not too sure to what.

Reply:
Let's do `convertAmplifyCategoryToCoreDataCategory`

Reply:
Dane! Thanks for a better name!